### PR TITLE
Feature/659/metric max nesting level

### LINKED ISSUE #659: New Metric: Nesting Level
# Feature request

## Description

Nesting Level describes the level of nested blocks of imperative source code such as nested if-clauses, while-clauses, repeat-until clauses etc.

As a user I want to be able to analyze the nesting level of given classes to gain a better understanding of the quality of the source code.

### PR DESCRIPTION
**Pull request by NearW**
# Added Metric for Maximum Nesting Level

closes #659

## Description
- Added a max_nesting_level metric to the SourceCodeParser
- Metric is calculated with the use of Visitors
- Refactored some functions in JavaAnalyzer and made it easier to add future metrics from visitors

## Definition of Done

A task/pull request will not be considered to be complete until all these items can be checked off.

- [x] **All** requirements mentioned in the issue are implemented
- [x] Does match the Code of Conduct and the Contribution file
- [x] Task has its own **GitHub issue** (something it is solving)
  - [x] Issue number is **included in the commit messages** for traceability
- [x] **Update the README.md** with any changes/additions made
- [x] **Update the CHANGELOG.md** with any changes/additions made
- [x] **Enough test coverage to ensure that uncovered changes do not break functionality**
- [x] **All tests pass**
- [x] **Descriptive pull request text**, answering:
  - What problem/issue are you fixing?
  - What does this PR implement and how?
- [x] **Assign your PR to someone** for a code review
  - This person _will be contacted **first**_ if a bug is introduced into `master`
- [x] **Manual testing** did not fail

### CODE REVIEW

File: analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        forwardNesting++
+
+        checkForChildrenAndUpdateMax()
+
+        super.visitDoWhileStatement(tree)
+
+        forwardNesting--
+    }
+
+    override fun visitLambdaExpression(lambdaExpressionTree: LambdaExpressionTree) {
```
Comment (NearW) [inline analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt]: @Richargh does a lambda-function increase the nesting level by one?

File: analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        currentNestingLevel++
+
+        potentiallyUpdateMaxNestingLevel()
+
+        super.visitWhileStatement(tree)
+
+        currentNestingLevel--
+    }
+
+    override fun visitSwitchExpression(tree: SwitchExpressionTree) {
```
[the remark this answers is not in the document]
Comment (NearW) [reply analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt]: Indentation doesn't mean that the nesting level increases though.
I think that a switch statement increases the nesting level by one as implemented and the individual cases only increase them, when they have if, for, etc. in it.

**Comment by NearW** [comment]
> Looks good, once we have confirmed the handling of lambda and switch expressions we can merge it 👍
> Just a thought: Should we include what metrics we can generate in the README of the SourceCodeParser?

That's a really good idea! I also wanted to add a some sort of wiki entry for future developers on how to use visitors to implemented new code metrics. I will update the README tomorrow.

File: analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (fatihi) [reply analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt]: What @NearW said. You could also just open code blocks for fun and that wouldn't increase the nesting level, so indentation is not an indicator.

File: analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt (hunk 1 above)
Comment (fatihi) [reply analysis/import/SourceCodeParser/src/main/kotlin/de/maibornwolff/codecharta/importer/sourcecodeparser/visitors/MaxNestingLevelVisitor.kt]: I couldn't find a formal definition, but I would say yes.

**Pull request merged**